### PR TITLE
feat(sdk): Enable AWS ALB authentication in KFP SDK Client

### PR DESCRIPTION
**Pull request by Jeffwan**
**Description of your changes:**

Currently, kfp client can not be used outside AWS EKS cluster. Application load balancer manages outside traffic and require authentication before traffic coming into mesh. This PR automates ALB authentication and get session cookie to authenticate KFP python client to Kubeflow cluster. 

This unblocks user to submit pipeline/run outside kubeflow cluster and user can integrate with their CI/CD solutions much easier. 

Cognito or OIDC behind ALB both can leverage this solution. 

Usage:

```
# Option 1. Pass username and password

from kfp import Client
HOSTNAME='https://kubeflow.platform.shanjiaxin.com/pipeline'
client = Client.create_cognito_authenticated(HOSTNAME, username='<your_username>', password='<your_passworkd>')
client.list_pipelines()

# Option 2. Interactively type your password without echoing.
client = Client.create_cognito_authenticated(HOSTNAME)


>> from kfp import Client
>>> HOSTNAME='https://kubeflow.platform.shanjiaxin.com/pipeline'
>>> client = Client.create_cognito_authenticated(HOSTNAME)

[WDM] - Current google-chrome version is 83.0.4103
INFO:WDM:Current google-chrome version is 83.0.4103
....
[/Users/shjiaxin/.config/kfp/chromedriver/drivers/chromedriver/mac64/83.0.4103.39]

Username: shjiaxin
Password:
>>> client.list_pipelines()
{'next_page_token': None,
 'pipelines': [{'created_at': datetime.datetime(2020, 7, 8, 7, 0, 13, tzinfo=tzutc()),
                'default_version': {'code_source_url': None,
                                    'created_at': datetime.datetime(2020, 7, 8, 7, 0, 13, tzinfo=tzutc()),
                                    'id': 'cd5251a2-18f6-40e6-ae7d-a7f1b9297928',

```

**Checklist:**
- [X] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.

- [x] Do you want this pull request (PR) cherry-picked into the current release branch?
    
    If yes, use one of the following options:
  
    * **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update.
    *  After this PR is merged, create a cherry-pick PR to add these changes to the release branch. (For more information about creating a cherry-pick PR, see the [Kubeflow Pipelines release guide](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#option--git-cherry-pick).)

Can we pick this PR to 1.0.0 release?

### CODE REVIEW

**Comment by Jeffwan** [comment]
@Ark-kun Is there a tool to format sdk projects?

**Comment by numerology** [comment]
> @Ark-kun Is there a tool to format sdk projects?

I think you can try [yapf](https://github.com/google/yapf). There is a config under `pipeline/.style.yapf` you can use.

**Comment by Jeffwan** [comment]
@numerology Thanks. I find your original PR https://github.com/kubeflow/pipelines/pull/2446 talking about coding styles. 

I tried to run the command however, its shows some hints on existing codes as well. I am wondering if CI has check on this and block the merge? Is this required step or optional?

**Comment by numerology** [comment]
> Is this required step or optional?

I think it's optional. The presubmit CI checks do not include this style check particularly.

**Comment by Jeffwan** [comment]
There's a python compatibility issue in the code. I will fix it today

**Comment by Jeffwan** [comment]
@numerology @Ark-kun @Bobgy 

I see the CI build with different python versions. Do we have to make sure python codes need to be compatible with 3.5 3.6 and 3.7? That means I can not use some library >=3.6?

**Comment by Bobgy** [comment]
@Jeffwan you are right.

/cc @Ark-kun 
Can explain more

**Comment by Ark-kun** [comment]
Thank you for working on new auth methods.
I think it's useful to have such functionality in KFP.

Several high-level comments:

1. We want to try and keep the interface of the `kfp.Client` platform-agnostic. It supports authentication using OAuth tokens. I wonder whether we can avoid adding AWS-specific and specialized (Cognito) methods to `Client`? If they're needed, then maybe they should go into `kfp/aws.py` and then communicate with `kfp.Client` using `token` or cookie.

2. The selenium and chrome dependencies might be pretty heavy. I'm not sure we should add them

3. I'm not sure using password in code and emulating the browser to authenticate using password conforms to AWS security guidelines. AFAIK, usually the password-based authentication happens outside the programs/libraries and the library only receives the token.

4. Authentication using OAuth and tokens might be preferred to authentication using passwords and cookies.

/cc @eterna2

**Comment by Ark-kun** [comment]
I've found this [article](https://lobster1234.github.io/2018/05/31/server-to-server-auth-with-amazon-cognito/) which describes a way to programmatically authenticate with Cognito and get a token that can be used with the standard Bearer authentication header.

You use ClientId and ClientSecret to send a token request and get back a JWT token that can be used for sending requests. 
```
$ curl -X POST \
  https://lobster1234.auth.us-east-1.amazoncognito.com/oauth2/token \
  -H 'authorization: Basic <base64-encoded (ClientId:ClientSecret)>' \
  -H 'content-type: application/x-www-form-urlencoded' \
  -d 'grant_type=client_credentials&scope=transactions%2Fpost'
```

**Comment by Bobgy** [comment]
> The selenium and chrome dependencies might be pretty heavy. I'm not sure we should add them

+1, this seems a big no to me too.

**Comment by eterna2** [comment]
I agree too. Selenium is too heavy. I can be wrong but I think ALB  only support cookies for now, and not bearer authorization.  

I would propose using the popular `requests` lib to do a headless authentication, we still can retrieve the session cookie this way. 

https://requests.readthedocs.io/en/latest/user/quickstart/#cookies

**Comment by Jeffwan** [comment]
> I've found this [article](https://lobster1234.github.io/2018/05/31/server-to-server-auth-with-amazon-cognito/) which describes a way to programmatically authenticate with Cognito and get a token that can be used with the standard Bearer authentication header.
> 
> You use ClientId and ClientSecret to send a token request and get back a JWT token that can be used for sending requests.
> 
> ```
> $ curl -X POST \
>   https://lobster1234.auth.us-east-1.amazoncognito.com/oauth2/token \
>   -H 'authorization: Basic <base64-encoded (ClientId:ClientSecret)>' \
>   -H 'content-type: application/x-www-form-urlencoded' \
>   -d 'grant_type=client_credentials&scope=transactions%2Fpost'
> ```

@Ark-kun @Bobgy   Thanks for the feedback. Using Cognito separately is a different flow. We are using application load balancer with Cognito in this case, I talked with cognito team internally and still need to double check with ALB team on this.  Currently, as I know, ALB doesn't accept intial Auth token from Cognito directly.  

Totally agree this is kind of heavy but seems there's no other ways to easily authenticate user

**Comment by Jeffwan** [comment]
@eterna2 do you know any ways to pass user info to ALB if we use `requests`?

**Comment by eterna2** [comment]
> @eterna2 do you know any ways to pass user info to ALB if we use `requests`?

```py
from requests.auth import HTTPBasicAuth

def get_auth_cookie(hostname: str, uid: str, pwd: str, schema: str = "https"):
  resp = requests.get("%s://%s/apikflogin" % (schema, hostname), auth=HTTPBasicAuth(uid, pwd))
  resp.raise_for_status()  # raise error if request is not ok
  return resp.cookies.items()[0]  # best if we know the regex for the cookie name instead of taking the first cookie

```

**Comment by eterna2** [comment]
If i am not wrong ALB checks the session cookie? So if u get the cookie from above, then to insert to all future requests, just put the cookie in the cookie jar. 

```py
import requests

jar = requests.cookies.RequestsCookieJar()
jar.set(<auth_cookie_name>, <auth_cookie_value>, domain='<domain>')
r = requests.get(url, cookies=jar)
```

**Comment by Jeffwan** [comment]
> kflogin

What's this? Anywhere I can find it? Is it basic auth? 

> If i am not wrong ALB checks the session cookie? 

Yeah, ALB expects cookie. Current challenge is there's no way to authenticate the user. User can go through UI to authenticate themselves and get cookie from browser for sure. But this is not enough to give end to end experiences for some scenarios like fully automate CI/CD.

**Comment by eterna2** [comment]
https://github.com/kubeflow/kubeflow/blob/0360d956a02970c0aae27bce591d34d903ce2d88/components/kflogin/src/login.js#L67

I am assuming this is the login frontend for kf. U can do the entire login process with requests headless. Just need to attach a cookie jar and figure out where are the authentication endpoint.

A cookie jar will store and manage the cookies from a regular http request w/o needing any frontend interaction.

**Comment by eterna2** [comment]
if u have an alb + cognitio endpoint for me to test, I can help with u on this PR.

**Comment by Jeffwan** [comment]
@eterna2 

> I am assuming this is the login frontend for kf. U can do the entire login process with requests headless. Just need to attach a cookie jar and figure out where are the authentication endpoint.

I think we are on the same page on the cookie part. The missing part is if we want to include login part in kfp or not.  We didn't use it kflogin, it's a Cognito page. 

> if u have an alb + cognitio endpoint for me to test, I can help with u on this PR.

I will send you username and password offline,

**Comment by eterna2** [comment]
@Jeffwan @Ark-kun @Bobgy 

Hey all, I just investigate this alb + cognito authentication process. It seems to me AWS did not make it easy for us to progammatically authenticate. There 2 solutions:

Solution 1:
Similar to what jeffwan did. We need to use pupeeter, selenium or some light-weight headless browser to progammatically login and get the token.

Solution 2: Preferred but has a problem
We can use boto3 or requests to make an oauth request to get the token. 

The only issue with this is UX and workflow. Because if u setup cognito to requires a client secret, this client secret is required to make an oauth request. And this means the data scientists will need to have access to the client secret before they can programmatically login cognito.

Alternatively, we can try setting up the cognito userpool w/o a client secret. But I am not sure if this is possible with ALB (I only use cognito alone w/o ALB). This means we probably need to update the docs and manifest/plugin on setting up the ALB + Cognito. 

What do u think @Jeffwan ?

**Comment by Bobgy** [comment]
For GCP, we also use OAuth and requires a client secret.

**Comment by Jeffwan** [comment]
@eterna2 

The limitation of Solution 1 is Cognito can connect to any IDP, current PR (dom structure) is based on native Cognito. If user change to Okta, Auth0, the login page will be different. That's my major concerns. 

Solution 2. I am not sure if the token from cognito can be used in ALB directly. If that's supported, I think that makes sense to use sdk to talk to cognito to exchange the token. 

Based on most user's feedback, I think using client_secret is ok. Most users need this in their CI/CD system, which means this will be used by robot account..

**Comment by NikeNano** [comment]
Are there any progress on this @Jeffwan, happy to help out if you need help?

**Comment by cloudbow** [comment]
This is a very important pull request for aws users as this is the only way we can use multi user isolation of pipelines programatically.

**Comment by DavidSpek** [comment]
Ping @PatrickXYS

**Comment by tkislan** [comment]
Is Authenticating to Kubeflow API on AWS somewhere on the roadmap anytime soon?
Because without this, this whole project is just worthless for production deployment, since you need trigger everything through UI only

**Comment by NikeNano** [comment]
@kubeflow/aws is this being worked on? Feels lika an important feature in order to support kubeflow on aws as pointed out https://github.com/kubeflow/pipelines/pull/4182#issuecomment-743223475. Would be happy to help out with the efforts as well.

**Comment by PatrickXYS** [comment]
I don't think anyone is doing this part, I'll be working on this when available.

**Comment by NikeNano** [comment]
Awesome @PatrickXYS, I am happy to help out if you need any help!

**Comment by jaystary** [comment]
Is this still active/worked on?

**Comment by Deseram** [comment]
Hi guys, is anyone looking into this? would be great if we can have this feat soon :pray:

**Comment by eterna2** [comment]
> Hi guys, is anyone looking into this? would be great if we can have this feat soon 🙏

I think no one at the moment. Because there was some challenges in programmatically authenticate against alb.

I haven't figure out what sort of checks cognito is doing when I try to mimic a login w/o actually having a headless browser (e.g. puppeteer or selenium).

Not sure if there is some x-csrf token or maybe even source up check.

Tldr when I try to programmatically login, cognito reject the oauth call.

**Comment by mtszkw** [comment]
Any progress on this? 😥 Is there any chance this can be done at all? @PatrickXYS

**Comment by DavidSpek** [comment]
For those getting hung up on this, [ArgoFlow for AWS](https://github.com/argoflow/argoflow-aws) is a distribution I started that deploys the Kubeflow (1.3) manifests using Argo CD and has integrations for AWS (such as S3 and RDS for pipelines). We re-implemented auth with Istio using OAuth2-Proxy which should make it easier to integrate with different OIDC providers. The current default setup is using an NLB in instance mode, where Istio handles ingress certificates on the gateways. With OAuth2-Proxy it shouldn't be difficult to set it up with Cognito. I'm not sure if that changes the SDK situation much, but given that the authentication is handles by Istio I expect the current auth mechanisms in the SDK would work.

File: sdk/python/kfp/_auth.py [hunk 1]
```diff
@@ -190,3 +209,161 @@ def id_token_from_refresh_token(client_id, client_secret, refresh_token, audienc
...
+    path = os.path.dirname(DEFAULT_CHROME_DRIVER_PATH)
+
+    if not os.path.exists(path):
+        os.makedirs(path)
+
+    z.extractall(path)
+    os.chmod(DEFAULT_CHROME_DRIVER_PATH, stat.S_IEXEC)
+
+
+def mkdir(path):
```
Comment (boarder7395) [inline sdk/python/kfp/_auth.py]: If defining this here it makes sense to use it on line 318 as well.

In both places I'm curious if there is a reason to avoid the following:
`os.makedirs(path, exist_ok=True)`

**Comment by rimolive** [comment]
Closing this PR. No activity for more than a year.

/close
